Use outputs to produce qiskit and qiskit-terra packages

wshanks opened the pull request:
* All of the old build configuration was moved under the qiskit-terra output
* The qiskit output installs the separate package under qiskit_pkg in the source code.
  - For the moment, the qiskit and qiskit-terra package versions and build numbers do not match, but with version 0.45, they should be synchronized.
  - The qiskit output could be noarch if it were possible for conda to support multiple outputs with an arch-specific subpackage and a noarch package that depends on it.
* The qiskit.algorithms test was removed as qiskit.algorithms  is moving to a separate package. See https://github.com/Qiskit/RFCs/pull/44.

<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

Closes #25

wshanks commented:
Hmm, currently I have done something bad that has pulled the entire site-packages into both `qiskit` and `qiskit-terra` packages and broken then pypy and macOS ARM cross-builds. I got the license files included correctly though :slightly_smiling_face:

At this point, the recipe is like two parallel recipes in one file, pulling different sources and building separately. The benefit of having them in the same recipe is that they will get published together. Otherwise, if they were separate feedstocks, the `qiskit` package would have to wait a couple hours after the `qiskit-terra` release was merged before its release could be merged. One option for tying the two packages closer together would be to use the GitHub source rather than the PyPI sources. Then there would be one common source directory for both packages at least. Maybe it's fine the way I have it besides the site-packages / cross-build issues. I just hope the grayskull bot updates can still work.

BastianZim commented on recipe/meta.yaml:
```diff
@@ -7,70 +7,113 @@ package:
 source:
-  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/qiskit-terra-{{ version }}.tar.gz
-  sha256: 1ea5981800377af5cc8beefd5ca62e18196b918058de8382a129abc921fa3c8b
+  - url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/qiskit-terra-{{ version }}.tar.gz
+    sha256: 1ea5981800377af5cc8beefd5ca62e18196b918058de8382a129abc921fa3c8b
+    folder: terra
+  - url: https://pypi.io/packages/source/q/qiskit/qiskit-0.44.1.tar.gz
+    sha256: 353c1aa2624c4044bd7013bd1e996a736c60934ebe323119c1f5b07527c6b05f
+    folder: qiskit
```
If it's coming from two different sources it might be easier to keep it in separate feedstocks for now, until it is fully merged.

wshanks replied:
Ultimately, all the source code comes from https://github.com/Qiskit/qiskit. Its publishing workflows produce separate sdists. If the only concern is that there are two sources, we could switch to the GitHub release source instead of the PyPI sources.

BastianZim commented:
> Could qiskit be noarch while qiskit-terra is arch-dependent? Or do subpackages need to all be arch-dependent.

Yes, that works and is done often.

> qiskit and qiskit-terra's versions do not match yet.

Then let's maybe keep it separate for now?

> I guess qiskit should not get the Rust thirdparty.yml file?

No, since it's noarch and there is no rust. qiskit will pull in qiskit-terra which will make sure that the license file is there.

> Will this break the helpful grayskull dependency updates?

Unfortunately, yes. Maybe another reason to stick with multiple feedstocks for now? Even if the other one is building an empty package?

BastianZim commented:
> Otherwise, if they were separate feedstocks, the qiskit package would have to wait a couple hours after the qiskit-terra release was merged before its release could be merged.

That can be managed with this option: https://conda-forge.org/docs/maintainer/conda_forge_yml.html#bot

```yaml
bot:
  # only open PRs if resulting environment is solvable, useful for tightly coupled packages
  check_solvable: true
```
And we could add an auto merge with Grayskull update which makes it all automated.

BastianZim commented:
Given everything above: I'm wondering if it's not easier to keep both feedstocks, add the bot options to the qiskit feedstock and just continue actively managing the qiskit-terra one with the other one being "dragged" along by Grayskull?

wshanks commented:
Thanks, @BastianZim! Before your comments, the tradeoffs for me were:

* Keep a single recipe. Give up grayskull updates. Add some complexity to the recipe (still not sure why the macOS cross-compile isn't working now). Keep the two packages publishing in sync.
* Split into two recipes. Keep grayskull updates and keep current terra recipe. The qiskit recipe should be simple. Packages would not be synchronized so every release terra would need to checked and merged. Then a couple hours later someone would need to come back and tag the qiskit recipe with automerge and restart its CI (since it would have failed due to the missing terra version the first time).

I was torn on the clunkiness of these two options (no more grayskull vs manually coming back to merge qiskit each release). `check_solvable` plus `automerge` seems like a good solution to the issue with separate packages, so I agree we should try that for now.

BastianZim commented:
Ok, all makes sense. I’ve merged the other one so we should be good to go now. 
I think, overall, this is probably the easier way. Even if it means having two feedstocks.

jakirkham commented:
With the multiple `outputs` case, the `pip` flags are not set by conda-build for some reason. This is a known bug: https://github.com/conda/conda-build/issues/3993

Would manually pass them all like so: https://github.com/conda-forge-admin/implicit-feedstock/commit/5362ea49a68efa60e1cbfad6ab1e284a4d990374

wshanks commented:
Thanks, @jakirkham. Bastian had pointed me at https://github.com/conda-forge/conda-forge.github.io/pull/1989 and from that I think I added all the flags already. I switched from `{{ PYTHON }}` to `python` because I saw that in some of the linked examples and it seemed like `{{ PYTHON }}` was not working. I see it in your linked example though, so I will try it again. Ultimately we will close this PR in favor of https://github.com/conda-forge/qiskit-feedstock, but I was curious if there were anything extra to add to https://github.com/conda-forge/conda-forge.github.io/pull/1989.

wshanks commented:
Hmm, with `{{ PYTHON }}` I still get an error:

```
INFO:conda_build.metadata:Attempting to finalize metadata for qiskit-terra
Error: Failed to render jinja template in /home/conda/recipe_root/meta.yaml:
'PYTHON' is undefined
```

I wonder what's different here from the `implicit` feedstock?

jaimergp commented on recipe/meta.yaml:
```diff
@@ -7,70 +7,113 @@ package:
-    - qiskit-toqm >=0.1.0
-    - qiskit-qasm3-import >=0.1.0
+outputs:
+  - name: qiskit-terra
```
Try changing the top-level `package.name` to something else, e.g. `qiskit-terra-split` so these two names do not collide.

wshanks replied:
Thanks. I just tried that, but I still see `PYTHON` as undefined in jinja rendering.

jaimergp replied:
I think that's because there's no top-level `requirements.host` entry. You could get away with `{{ PREFIX }}/bin/python` or just `python`, though. Does that not work?

wshanks replied:
Oh, sorry, I see what the problem is now. When I use `python` I get `Bad CPU type in executable`, which I thought meant the cross-build was using the wrong Python. Now I see that that error comes just after it starts building the second noarch Python wrapper. It actually builds the arch-specific package okay. So the trick is just that I need to set up the wrapper package to be noarch or set up cross-Python for it.

wshanks replied:
@jaimergp I tried making the second package `noarch: python` in [e863332](https://github.com/conda-forge/qiskit-terra-feedstock/pull/28/commits/e863332ce3aa6f6d2c668d572834ca82546a24ba) but I still got the Bad CPU error. Should something like this be possible?

jaimergp replied:
Looks like it works?

jaimergp commented on recipe/meta.yaml:
```diff
@@ -3,74 +3,122 @@
+      script: python -m pip install ./qiskit -vv --no-deps --no-index --no-build-isolation --ignore-installed --cache-dir pip_cache
+    requirements:
+      build:
+        - python                                 # [build_platform != target_platform]
```
Do you need cross-compilation for `noarch`? 🤔

wshanks replied:
I got tunnel vision from the failures and trying to get everything green. I think your next comment about skipping all builds except one is the real answer here.

jaimergp commented on recipe/meta.yaml:
```diff
@@ -3,74 +3,122 @@
+  - name: qiskit
+    version: 0.44.1
+    build:
+      noarch: python
```
You should probably build this only once by adding the relevant `skip` lines. Otherwise you'll get collisions.

I'll keep it for Linux x64 only. It should work on the others.

wshanks replied:
Thanks! I went down a rabbit hole trying to figure out why some things were failing, but, now that I understand it was just that even a noarch package build for osx-arm needs cross-python, I see that the solution is just to let the feedstock build all the arch-specific versions, but then skip the noarch version for all variants except one.

wshanks commented:
Everything makes sense now, thanks @jaimergp! I will close this because we decided to keep this feedstock as just for qiskit-terra and open a separate dependent qiskit-feedstock for the wrapper that will be pulled along by `bot: check_solvable: true` and `grayskull-update`.
